catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,447 @@
+        currentSourceKeyAndSourceMap.entrySet()) {
+      final Source source = sourceKeyAndSourceEntry.getValue();
+      final String sourceId = source.getId();
+      final SourceKey sourceKey = sourceKeyAndSourceEntry.getKey();
 
-    handle = scheduler.scheduleAtFixedRate(runner, INITIAL_DELAY, interval, TimeUnit.MINUTES);
+      boolean sourceIsNotInCache = cache.getIfPresent(sourceKey) == null;
+      if (sourceIsNotInCache) {
+        final SourceStatus initialSourceStatus = SourceStatus.UNKNOWN;
+        LOGGER.debug(
+            "Found a new SourceKey for source id={}. This is caused when a source is created or when a source is modified in a way that cause its SourceKey to be different. Adding this source to the cache with an initial status of {}",
+            sourceId,
+            initialSourceStatus);
+        cache.put(sourceKey, initialSourceStatus);
+      }
+
+      LOGGER.trace("Starting a process to check the availability of source id={}", sourceId);
+      final ListenableFuture sourceIsAvailableFuture =
+          Futures.withTimeout(
+              sourceIsAvailableThreadPool.submit(
+                  () -> {
+                    try {
+                      final SourceStatus sourceStatus =
+                          source.isAvailable() ? SourceStatus.AVAILABLE : SourceStatus.UNAVAILABLE;
+
+                      // TODO Would it be worthwhile to log something at info when the source status
+                      // changes?
+                      // TODO Is there a simpler way to cache the source status as soon as we know
+                      // it (e.g. handling futures by first completed)?
+                      LOGGER.trace(
+                          "Found that the source id={} is {}. Caching the new source status",
+                          sourceId,
+                          sourceStatus);
+                      cache.put(sourceKey, sourceStatus);
+                    } catch (RuntimeException e) {
+                      final SourceStatus sourceStatus = SourceStatus.EXCEPTION;
+                      LOGGER.debug(
+                          "RuntimeException checking the availability of source id={}. Caching source status as {}",
+                          sourceId,
+                          sourceStatus,
+                          e);
+                      cache.put(sourceKey, sourceStatus);
+                    }
+                  }),
+              timeout,
+              timeoutTimeUnit,
+              sourceIsAvailableTimeoutWatcherThreadPool);
+
+      futureAndSourceKeyMap.put(sourceIsAvailableFuture, sourceKey);
+    }
+    LOGGER.trace("Successfully submitted availability checks for all of the sources");
+
+    for (final Entry<Future, SourceKey> futureAndSourceKeyEntry :
+        futureAndSourceKeyMap.entrySet()) {
+      final Future future = futureAndSourceKeyEntry.getKey();
+      final SourceKey sourceKey = futureAndSourceKeyEntry.getValue();
+
+      try {
+        future.get();
```

Comment:
❓ It is possible I'm reading this wrong (this method is huge!), but won't this potentially cause all of the sources to sequentially wait for long running sources to return status? Is that what we want to do here? We could do each of these checks in their own thread, conversely.

Reply:
I know it's big. I wanted to get feedback before pulling out some helper methods. It fails the SonarLint complexity check right now and I left a TODO code comment, so it won't go unfixed:+1:

Your concern is valid, and it was a design decision that we discussed, but I think it's addressed here, albeit maybe not in the best/simplest way.

The SourceStatues will immediately be updated in the cache as soon as the availability check completes (AVAILABLE, UNAVAILABLE, and EXCEPTION) for availability checks that finish more quickly than 1 poll interval. The runnable where we create the `Futures.withTimeout` is where we check the availability and immediately cache it. Then, we loop through all of the "timeout" futures one by one. If the future has already finished successfully, we don't cache anything because the new SourceStatus was already cached. If the future failed because of a timeout, then we cache the SourceStatus as TIMEOUT. So, the only time that we will wait for long running sources is when we serially loop through the "timeout" futures and cache the TIMEOUT SoureStatuses, which is expected. I'll add a code comment here to clarify what's happening in this second `for` loop.

Any suggestions to improving how this is accomplished would be great. I did investigate using [Futures#inCompletionOrder](https://google.github.io/guava/releases/20.0/api/docs/com/google/common/util/concurrent/Futures.html#inCompletionOrder-java.lang.Iterable-), using a [CompletionService](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletionService.html), and managing the thread timeouts manually instead of using [Futures#withTimeout](https://google.github.io/guava/releases/20.0/api/docs/com/google/common/util/concurrent/Futures.html#withTimeout-com.google.common.util.concurrent.ListenableFuture-long-java.util.concurrent.TimeUnit-java.util.concurrent.ScheduledExecutorService-). I ended up settling on the solution that I have now, but there is room for improvement.

Reply:
I added the code comment:
```java
// Now go through each of the Futures one-by-one. If the Future completes successfully without
// throwing an Exception, the new SourceStatus was already successfully cached. If an Exception
// is thrown, check if it is a TimeoutException. In that case, cache the SourceStatus as
// TIMEOUT. Otherwise, there is some other unexpected (thread-related) error so do not update
// the cache.
```